frame/program.py

```diff
@@ -63,25 +54,40 @@ def wait(self, timeout=default_wait_timeout) -> None:
                 message += ' closed with exit code ' + str(self.process.returncode)
             raise RuntimeError(message)
 
-    def kill(self, timeout=default_term_timeout) -> None:
-        if self.process.returncode == None:
+    async def kill(self, timeout=default_term_timeout) -> None:
+        assert self.process
+        if not self.waited_for:
             os.killpg(self.process.pid, signal.SIGTERM)
             try:
-                self.wait(timeout=timeout)
-            except subprocess.TimeoutExpired:
+                await asyncio.wait_for(self.wait(), timeout=timeout)
+            except asyncio.exceptions.TimeoutError:
                 pass
-        if self.process.returncode == None:
+        if not self.waited_for:
             os.killpg(self.process.pid, signal.SIGKILL)
             self.killed = True
-            self.wait(timeout=timeout)
+            try:
+                await asyncio.wait_for(self.wait(), timeout=1)
+            except asyncio.exceptions.TimeoutError:
+                raise RuntimeError('failed to kill ' + self.name)
```

Comment:
Got lost here. When will `SIGKILL` ever be sent? If we didn't wait, yet, then we'll `.wait()`, so `.waited_for` will be set `True`, so `SIGKILL` will be skipped over; if we _did_ wait, it's all NOOP?

Reply:
OK the Timeout will come from the `asyncio.wait_for`, and only if the `.communicate` in `wait()` didn't finish.